_posts/2021-06-09-fed-talk-s01e03-routing.md

```diff
@@ -0,0 +1,513 @@
+  const { push } = useHistory();
+
+  const onLinkNavigation = (url: string) => {
+    push(url);
+    setOpenMenu(false);
+  }
+
+  return (
+    <List>
+      {menuItems.map(({ label, url }) => (
+        <ListItem button key={label} onClick={() => onLinkNavigation(url)}>
+          <ListItemText primary={label} />
+        </ListItem>
+      ))}
+    </List>
+  )
+}
+
+/* ...Rest of code */
+```
+
+**Notes:**
+* We moved the `menuItems` outside the component, this is simply to initialize the menuItems once and refer to it there after.
+* We declare the use of the `History` hook and explicitly require its `push` function for future use.
+* We then created a function `onLinkNavigation` to manage the users click event. Upon clicking we instruct the App to push the new navigation URL into the browsers History queue; then we hide the menu.
+
+Here's what this new change looks like:
+
+![Screen captured gif illustrating a user navigating through the 3 UI pages via the left menu links](/img/fed-talk/s01e03/left-nav-animation.gif){:width="350"}
+
+:warning:
+**Hang on, this implementation has flaws!**
+
+Even though this functionally works, it's unfortunately not accessible!
+MUI Have realized this is a problem and have provided a way for us to integrate 3rd party components such as `react-router-dom` `Link` component; which would ultimately render our `ListItem` component as an anchor tag, with a href value.
+
+Let's make the changes:
+
+```tsx
+// src/components/MainMenu/index.tsx
+
+import React from "react";
+import { Link } from "react-router-dom";
+
+import { Drawer, List, ListItem, ListItemText } from "@material-ui/core";
+
+const menuItems = [
+  { label: 'Buttons', url: '/ui/buttons' },
+  { label: 'Colour Palette', url: '/ui/colour-palette' },
+  { label: 'Typogaphies', url: '/ui/typographies' },
+]
+
+function MenuItems({ setOpenMenu }: { setOpenMenu: React.Dispatch<React.SetStateAction<boolean>> }) {
+  return (
+    <List>
+      {menuItems.map(({ label, url }) => (
+        <ListItem
+          button
+          component={Link}
+          key={label}
+          onClick={() => setOpenMenu(false)}
+          to={url}
+        >
+          <ListItemText primary={label} />
+        </ListItem>
+      ))}
+    </List>
+  )
+}
+
+/* ...Rest of code */
+```
+
+**Notes:**
+* We've imported the `Link` component from `react-router-dom` and then passed it through to the `ListItem` "component" property. This then extends the TypeScript definition of `ListItem` with the types of `Link`, making the "to" property available.
+* We then removed the need to include the `History` hooks as we've passed the menuItem's url value into the "to" property.
+* We update the "onClick" property to collapse the main menu there after.
+
+:champagne:
+Those links are now accessible!
+
+---
+
+## Parameterized Routes
+
+Depending on your App's architecture and the data in which it needs to process, there will be a time where you need to configure URL parameters.
```

Comment:
I wouldn't say this is specific to React Router and I would probably mention the difference between **query parameters** and **path parameters**.

https://medium.com/@fullsour/when-should-you-use-path-variable-and-query-parameter-a346790e8a6d

React Router does still allow the use of query parameters and I believe they can still be useful and readable.

Reply:
This puts it really nicely from the article above:

_If you want to identify a resource, you should use Path Variable. But if you want to sort or filter items, then you should use query parameter._

Reply:
Yeah in this area I'm focusing on React-Router's "params" prop (path parameters). But you raise a good point, I haven't mentioned anything about the "search" prop (query params). 
<img width="442" alt="Screen Shot 2021-06-11 at 10 29 12 am" src="https://user-images.githubusercontent.com/25266506/121622175-2177a780-caa0-11eb-9956-9273ed651a13.png">